### ISSUE TITLE
feat(beam): path-dependent grammar and schema-constrained beam search for structured tool extraction

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12411]
# feat(beam): path-dependent grammar and schema-constrained beam search for structured tool extraction

```routing
lane: internal/beam
paths: internal/beam/constrained.go, internal/model/constraint.go, internal/beam/constrained_test.go
expected_steps: 4
```

**Lane:** internal/beam  
**Work Unit:** leaf  
**Generation:** gen/now  
**Priority:** P2  

## Parent context
Extends #12407 (`feat(beam): zero-allocation beam search engine spine`). Integrates beam search with `internal/model/constraint.go` (`DecodeConstraint`, `LogitMask`, `LogitBias`) and `internal/guideddecode` to enforce strict JSON schemas and tool-call grammars across multiple search hypotheses.

## Current state
`fak`'s constrained decoding (`internal/model/constraint.go`) evaluates `LogitMask.MaskLogits(history, logits)` under the assumption of a single sequence history. In beam search, each candidate beam $b$ has generated a distinct prefix $H_b$. For example, Beam 0 might have emitted `{"name":"bash","args":{"command":"` while Beam 1 emitted `{"name":"read","args":{"file":`. Evaluating a single global constraint cannot support branching path-dependent grammar states.

## Why this is next
Structured tool calling is the primary operational domain where beam search provides massive value over greedy decoding: greedy decode easily falls into syntactic traps or invalid schema transitions, whereas constrained beam search explores valid grammar completions and recovers from near-miss dead ends. Enforcing an Adjudicator Fence ensures tool calls are never executed during search.

## Working spine
Track per-beam history prefix $H_b$ -> apply `DecodeConstraint.MaskLogits(H_b, logits[b])` per beam before candidate expansion -> mask invalid grammar transitions to $-\infty$ -> allow only schema-compliant tokens to enter the top-$2B$ candidate pool -> enforce the Adjudicator Fence (zero tool execution side-effects during search) -> verify 100% schema compliance on the winning output.

## Core through-line
1. Implement `ConstrainedBeamSearch` in `internal/beam/constrained.go` accepting `*model.DecodeConstraint`.
2. For each active beam $b \in [0, B-1]$ at step $t$, resolve its history $H_b$ from the arena backpointers.
3. Call `constraint.MaskLogits(H_b, logits[b])` in-place on beam $b$'s logits prior to candidate selection, setting invalid grammar transitions to $-\infty$.
4. Ensure invalid tokens naturally drop out of the top-$2B$ candidate pool, guaranteeing that all surviving hypotheses strictly conform to the active schema or grammar DFA/PDA.
5. Implement the Adjudicator Fence: guarantee that tentative tool calls emitted on speculative beams are never dispatched to external tools or system execution environments until search completes and the winning beam is selected.
6. Author tests in `internal/beam/constrained_test.go` verifying that complex JSON schemas (nested objects, required keys, enum values) are generated with 100% validity across diverse prompt seeds.

## In scope
- Path-dependent per-beam logit masking.
- Integration with `model.DecodeConstraint`, `LogitBias`, and `LogitMask`.
- Adjudicator Fence preventing tool side-effects during multi-path search.
- Unit tests verifying JSON schema validation over challenging synthetic and real-world prompts.

## Out of scope
- Grammar compilation logic (owned by `internal/grammar` and `internal/guideddecode`).
- Multi-turn conversation state rollback (search is bounded to the single turn generation).

## Gold-plating boundary
Strictly excluded:
- Dynamic schema generation or LLM-synthesized regex grammars.
- Non-deterministic grammar backtracking.

## Verifiable Witness
```bash
go test -v ./internal/beam -run TestConstrainedBeamJSON
go test -v ./internal/beam -run TestAdjudicatorFenceDuringSearch
```

## Done condition
`internal/beam` decodes structured JSON matching complex schemas across all surviving beams, invalid transitions are masked to $-\infty$, zero side-effect tool calls are dispatched during search, and the winner satisfies schema validation.

## Scoped Acceptance Criteria
- [ ] 1. Each beam evaluates its specific prefix history $H_b$ against `LogitMask`. `[SW-VERIFIED]`
- [ ] 2. Invalid grammar tokens receive $-\infty$ logits and never enter the candidate survivor pool. `[SW-VERIFIED]`
- [ ] 3. Adjudicator Fence invariant is verified: 0 tool executions occur during active beam expansion. `[SW-VERIFIED]`
- [ ] 4. All completed beams parse as valid JSON conforming to the requested schema. `[SW-VERIFIED]`

## Acceptance gate
`go test -v ./internal/beam -run TestConstrainedBeam` exits 0 with 0 failures.

## Closure binding
Close via an explicit-path trunk commit landing `internal/beam/constrained.go` and associated test witnesses.

## Likely files
- `internal/beam/constrained.go`
- `internal/model/constraint.go`
- `internal/beam/constrained_test.go`
- `internal/adjudicator/fence.go`
